packages/react-reconciler/src/ReactUpdateQueue.js

```diff
@@ -117,27 +117,21 @@ export type Update<State> = {
   payload: any,
   callback: (() => mixed) | null,
 
-  next: Update<State> | null,
-  nextEffect: Update<State> | null,
+  next: Update<State>,
 
   //DEV only
   priority?: ReactPriorityLevel,
 };
 
+type SharedQueue<State> = {
+  pending: Update<State> | null,
+};
+
 export type UpdateQueue<State> = {
   baseState: State,
-
-  firstUpdate: Update<State> | null,
-  lastUpdate: Update<State> | null,
-
-  firstCapturedUpdate: Update<State> | null,
-  lastCapturedUpdate: Update<State> | null,
-
-  firstEffect: Update<State> | null,
-  lastEffect: Update<State> | null,
-
-  firstCapturedEffect: Update<State> | null,
-  lastCapturedEffect: Update<State> | null,
+  baseQueue: Update<State> | null,
+  shared: SharedQueue<State>,
+  effects: Array<Update<State>> | null,
```

Comment:
Why is this an array as opposed to a linked list like the hooks form? Harder to see how they correlate.

I was considering a refactor to arrays but with a different take (inline fields instead of objects) . However, I wanted to do that as a separate refactor.

Reply:
Idk I just chose one :D

This felt like a different trade off from the hooks one because these effects are relatively rare. Only when you pass a callback to `setState`. It's more like the `emitEffect` API.

Reply:
So it didn't seem worth putting a `nextEffect` field on the update type